### DISCUSSION TITLE
Unable to boot Jetson Nano Devkit r32.5.0 branch tegra-demo-distro

### DISCUSSION THREAD

**Discussion by pgiangrossi** [OE4T/meta-tegra#671] [Q&A] [answered]
I tried to deploy the demo-image-base using commit 60e217fb92bb99cb26403a2737b87662ac93c50a (HEAD -> dunfell-l4t-r32.5.0, origin/dunfell-l4t-r32.5.0).

I built using:
```sh
cd tegra-demo-distro
. ./setup-env --machine jetson-nano-devkit
MACHINE=jetson-nano-devkit bitbake demo-image-base
```

I then flashed the SD card on my Linux host via:
```sh
cd tmp/deploy/images/jetson-nano-devkit
tar -xvf demo-image-base-jetson-nano-devkit.tegraflash.tar.gz
./dosdcard.sh
sudo dd if=demo-image-base.sdcard of=/dev/sdd bs=10M status=progress
```

I ran into the following problem:
-  mount failure at startup (/dev/mmcblk0p not found, my sdcard had /dev/mmcblk0p1). I fixed this by adding distro_bootpart ?= "1" to conf/machine/jetson-nano-devkit.conf)

With this fix, I was able to build, deploy and bring up the Jetson Nano devkit.

I wonder if I am doing something wrong, or using this repo in the wrong way?

**Comment by ichergui** [comment, marked as answer]
Hey @pgiangrossi

I think you missed some steps in the build image process.

You don't need the cmd `$ sudo dd if=demo-image-base.sdcard of=/dev/sdd bs=10M status=progress` to flash the SDcard because the script `dosdcard.sh` already done this for you. 

* Here the instructions:

```
$ git clone https://github.com/OE4T/tegra-demo-distro.git
$ cd tegra-demo-distro
$ git checkout origin/dunfell-l4t-r32.5.0 -b dunfell-l4t-r32.5.0
$ git submodule update --init --recursive 
$ unset NVIDIA_DEVNET_MIRROR                                    
$ export NVIDIA_DEVNET_MIRROR="/PATH/TO/JETPACK_4.5/nvidia/sdkm_downloads"
$ . ./setup-env --machine jetson-nano-devkit --distro tegrademo build_jetson_nano
$ bitbake demo-image-base
```

* Go to the deploy directory and flash the device:
```
$ cd tmp/deploy/images/jetson-nano-devkit
$ tar -xvf demo-image-base-jetson-nano-devkit.tegraflash.tar.gz
```
In case of updating the `SPI flash` you need to use the script `doflash.sh` instead or You can do it in different way.

* Here is the steps
1- Flash the SDcard via the script `dosdcard.sh`
2- Boot the device and stop the boot process to have  the `u-boot` prompt  
3- Execute the following commands:
```
# env default -a
# saveenv
# reset
 ```

**Reply by pgiangrossi** [reply]
Thanks for the quick reply! I'll try the steps you list, I didn't know dosdcard.sh could flash the SD card. I guess I have to give some arguments for this, but I can check that.

I see c-boot in the image that I want to deploy. I haven't gotten U-Boot deployed yet, but I plan on it. Will the process you list above work with c-boot?

**Reply by ichergui** [reply]
Please refer to the [link](https://github.com/OE4T/meta-tegra/wiki/Flashing-the-Jetson-Dev-Kit#writing-an-sdcard) for flashing the device.

For your information, `Nvidia Jetson Nano DevKit` have `u-boot` as default bootloader. Are you configuring the default bootloader to be only `cboot` ?

Here is the traces from the serial UART
```
U-Boot 2021.01 (Feb 28 2021 - 13:04:45 +0000)

SoC: tegra210
Model: NVIDIA Jetson Nano Developer Kit
Board: NVIDIA P3450-0002
DRAM:  4 GiB
MMC:   sdhci@700b0000: 1, sdhci@700b0600: 0
Loading Environment from MMC... *** Warning - bad CRC, using default environment

In:    serial
Out:   serial
Err:   serial
Net:   No ethernet found.
Hit any key to stop autoboot:  0 
switch to partitions #0, OK
mmc0(part 0) is current device
Scanning mmc 0:1...
Found /boot/extlinux/extlinux.conf
Retrieving file: /boot/extlinux/extlinux.conf
224 bytes read in 14 ms (15.6 KiB/s)
1:      OE4Tegra Demonstration Distro
Retrieving file: /boot/extlinux/../initrd
2108826 bytes read in 65 ms (30.9 MiB/s)
Retrieving file: /boot/extlinux/../Image
37210120 bytes read in 872 ms (40.7 MiB/s)
append: tegraid=21.1.2.0.0 ddr_die=4096M@2048M section=512M memtype=0 vpr_resize usb_port_owner_info=0 lane_owner_info=0 emc_max_dvfs=0 touch_id=0@63 video=tegrafb no_console_suspend=1 console=ttyS0,115200n8 debu 
Moving Image from 0x84000000 to 0x84080000, end=8649b000
## Flattened Device Tree blob at 83100000
   Booting using the fdt blob at 0x83100000
ERROR: reserving fdt memory region failed (addr=0 size=0)
ERROR: reserving fdt memory region failed (addr=0 size=0)
   Using Device Tree in place at 0000000083100000, end 000000008317e6e7
copying carveout for /host1x@50000000/dc@54200000...
copying carveout for /host1x@50000000/dc@54240000...

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0
[    0.000000] Linux version 4.9.201-l4t-r32.5+gc68230c5b46a (oe-user@oe-host) (gcc version 8.3.0 (GCC) ) #1 SMP PREEMPT Thu Mar 4 21:10:03 UTC 2021
[    0.000000] Boot CPU: AArch64 Processor [411fd071]

```
Please check the Nvidia [documentation](https://docs.nvidia.com/jetson/l4t/index.html#page/Tegra%20Linux%20Driver%20Package%20Development%20Guide/bootflow_jetson_nano.html#wwconnect_header) about the `Jetson Nano boot flow`

**Reply by pgiangrossi** [reply]
OK, great. I see the command for dosdcard in the link you provided. Thanks!

I am using the git clone for tegra-demo-distro as is. No modifications. I haven't gotten the console working yet (I have tried the two serial ports that come to headers on the Nano dev kit board, no luck), so I can't verify what bootloader is on the board. But I do see c-boot.bin in the tar.gz file that I create with bitbake (and no u-boot binaries are there).

**Reply by ichergui** [reply]
> OK, great. I see the command for dosdcard in the link you provided. Thanks!
> 
> I am using the git clone for tegra-demo-distro as is. No modifications. I haven't gotten the console working yet (I have tried the two serial ports that come to headers on the Nano dev kit board, no luck), so I can't verify what bootloader is on the board. But I do see c-boot.bin in the tar.gz file that I create with bitbake (and no u-boot binaries are there).

You should see the `boot.img` binary which is equivalent of the `u-boot` binary.

**Reply by ichergui** [reply]
The serial debug console is available on the J50 header on the Jetson Nano.
```
Jetson Nano J50 Pin 4 (TXD) <--> Cable RXD (Yellow Wire)
Jetson Nano J50 Pin 3 (RXD) <--> Cable TXD (Orange Wire)
```
Please the image below
![IMG_9120](https://user-images.githubusercontent.com/15719516/115277309-06da1f80-a13c-11eb-8503-9314cac6330e.jpg)

**Reply by pgiangrossi** [reply]
I can now see the console on the serial port. I needed a 3.3V UART/USB and not a 5V. I was able to follow the instructions you listed in your first reply, even the U-boot environment settings. 

When I boot up with the serial connected, I do not see any mounting errors. However, I cannot login via the console. I do not see any characters echoed in the minicom window (but I could type in characters at the U-Boot prompt). I could not connect via ethernet either. Console log is attached:

[console.log](https://github.com/OE4T/meta-tegra/files/6341594/console.log)


If I boot up with a monitor connected to the HDMI, I still see the mounting problem that I described in the initial description.

**Reply by pgiangrossi** [reply]
printenv cbootargs 
cbootargs=tegraid=21.1.2.0.0 ddr_die=4096M@2048M section=512M memtype=0 vpr_resize usb_port_owner_info=0 lane_owner_info=0 emc_max_dvfs=0 touch_id=0@63 video=tegrafb no_cons
ole_suspend=1 console=ttyS0,115200n8 debug_uartport=lsport,4 earlyprintk=uart8250-32bit,0x70006000 maxcpus=4 usbcore.old_scheme_first=1 lp0_vec=0x1000@0xff780000 core_edp_mv
=1125 core_edp_ma=4000 gpt  earlycon=uart8250,mmio32,0x70006000  root=/dev/mmcblk0p1 rw rootwait rootfstype=ext4 console=ttyS0,115200n8 console=tty0 fbcon=map:0 net.ifnames=
0

**Reply by ichergui** [reply]
Hey @pgiangrossi 
The bootargs looks OK to me. 
Are you using only `Micro USB` as a power supply for the `Jetson nano dev kit` or `DC Barrel Jack` ?
I recommend to use the `DC Barrel Jack` as power supply for the `Jetson Nano`. Please refer to this [link](https://forums.developer.nvidia.com/t/power-supply-considerations-for-jetson-nano-developer-kit/71637).

Here is my setup. 

![IMG_9130](https://user-images.githubusercontent.com/15719516/115388082-62a3b780-a1d3-11eb-9b39-b0319df18be9.jpg)

**Reply by pgiangrossi** [reply]
I am using the Micro USB as power supply. When I apply the patch (distro_bootpart ?= 1) as described above, with the steps that you list, I get a login screen on the console and I can login to root. Without the patch, fails as indicated above.

I will try without the patch and the barrel jack as power source, and come back to you with results.

**Reply by ichergui** [reply]
From your serial UART the mounting of the rootfs is not happening.
Here is my traces from the serial uart as a reference:

```
[    5.089240] mmc0: hw tuning done ...
[    5.089272] mmc0: new ultra high speed SDR104 SDHC card at address aaaa
[    5.089673] mmcblk0: mmc0:aaaa SE32G 29.7 GiB 
[    5.093023]  mmcblk0: p1
[    5.102751] usb 2-1: usb_suspend_both: status 0
[    5.102896] usb usb2: usb_suspend_both: status 0
[    5.726708] usb 1-2: New USB device strings: Mfr=1, Product=2, SerialNumber=0
[    5.726712] usb 1-2: Product: 4-Port USB 2.1 Hub
[    5.726716] usb 1-2: Manufacturer: Generic
[    5.738589] hub 1-2:1.0: USB hub found
[    5.742617] hub 1-2:1.0: 4 ports detected
[    5.754394] Freeing unused kernel memory: 10624K
[    5.770040] EXT4-fs (mmcblk0p1): couldn't mount as ext3 due to feature incompatibilities
[    5.780215] EXT4-fs (mmcblk0p1): couldn't mount as ext2 due to feature incompatibilities
[    5.797362] EXT4-fs (mmcblk0p1): mounted filesystem with ordered data mode. Opts: (null)

```

You should not apply any patch and it should be working fine. 
As I mentioned above please use the script `doflash.sh` to flash your device.
I will flash my device because I'm using the master branch then share with you the traces from the Serial UART.
Do you remember which branch was flashed your device ? 
**For example:** 
* If your device was using the `dunfell-l4t-r32.4.3` before the `dunfell-l4t-r32.5.0`. In this case you need to flash the device via `doflash.sh`
Please let me know if it is not clear for you.

**Reply by ichergui** [reply]
Hello @pgiangrossi 

I flashed my device with the same branch that you are using. All good here.

The attached file contains the complete logs from the serial UART.

[jetson_nano_devkit_serial_uart.log](https://github.com/OE4T/meta-tegra/files/6344685/jetson_nano_devkit_serial_uart.log)

**Reply by pgiangrossi** [reply]
Hi @ichergui 

Using doflash.sh worked perfectly! 

I do not know what was on the board before I started flashing it, but it was probably the NVIDIA distro that you can download directly from their webiste (ie, it wasn't built locally using yocto).

My understanding is that the doflash.sh script updates the SPI flash as well as the SD card. Is that correct? So I guess the SPI flash needed updating. What is stored there vs. the SD card?

Thanks so much for your help!

/Paul

**Reply by ichergui** [reply]
Hey @pgiangrossi 
Glad to hear that.
All boot partitions are stored in the `SPI flash` that's why we need to update them. 
Yeah, I confirm that the script `doflash.sh` updates the `APP` (rootfs) partition as well.
